### PR TITLE
expanded test environment documentation

### LINKED ISSUE #3286: add full picture documentation for venv interaction in the compiler
See [slack discussion](https://inmanta.slack.com/archives/CKRF0C8R3/p1632145680214500).

### PR DESCRIPTION
**Pull request by sanderr**
# Description

As promised I've added a bit more documentation about the use of venvs in testing. Let me know if anything is not clear. I've spent enough time working with these venvs that I'm not sure I can accurately pinpoint the points of confusion.
@bartv I'm adding you as a reviewer because you haven't worked with these new environments yet (I think) but you do know about the design, so you should be able to understand it. If you don't, I probably need to clear some things up.

closes #3286

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] ~~Type annotations are present~~
- [x] ~~Code is clear and sufficiently documented~~
- [x] ~~No (preventable) type errors (check using make mypy or make mypy-diff)~~
- [x] ~~Sufficient test cases (reproduces the bug/tests the requested feature)~~
- [x] ~~Correct, in line with design~~
- [x] E~~nd user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -30,6 +30,31 @@
...
+to make sure those are installed in a test environment. This means patching the inmanta.env.process_env to be an interface
+to a test environment instead of the outer environment.
+The following fixtures manage test environments:
+- snippetcompiler_clean: activates the Project's compiler venv and patches inmanta.env.process_env to use this same venv
+    as if it were the outer venv. The activation and patch is applied when compiling the first snippet.
+- snippetcompiler: same as snippetcompiler_clean but the compiler venv is shared among all functions using the
+    fixture.
+- tmpvenv: provides a decoupled test environment (does not inherit from the outer environment) but does not activate it and
+    does not patch inmanta.env.process_env.
+- tmpvenv: provides a decoupled test environment, activates it and patches inmanta.env.process_env to point to this
```
Comment (arnaudsjs) [inline tests/conftest.py:51]:
```suggestion
- tmpvenv_active: provides a decoupled test environment, activates it and patches inmanta.env.process_env to point to this
```

**Review by bartv — APPROVED** [review]

File: tests/conftest.py [hunk 2]
```diff
@@ -30,6 +30,31 @@
...
+
+When running the tests we don't want to make changes to the outer environment. So for tests that install Python packages we need
+to make sure those are installed in a test environment. This means patching the inmanta.env.process_env to be an interface
+to a test environment instead of the outer environment.
+The following fixtures manage test environments:
+- snippetcompiler_clean: activates the Project's compiler venv and patches inmanta.env.process_env to use this same venv
+    as if it were the outer venv. The activation and patch is applied when compiling the first snippet.
+- snippetcompiler: same as snippetcompiler_clean but the compiler venv is shared among all functions using the
+    fixture.
+- tmpvenv: provides a decoupled test environment (does not inherit from the outer environment) but does not activate it and
```
Comment (arnaudsjs) [inline tests/conftest.py:49]: The docstring of the `tmpvenv` fixture mentions that it is activated.

File: tests/conftest.py [hunk 3]
```diff
@@ -30,6 +30,31 @@
...
+- inmanta.module.Project.virtualenv: inmanta.env.VirtualEnv -> compiler venv. Used by the compiler to install v1 module
+    requirements in. Inherits from the outer venv.
+
+When running the tests we don't want to make changes to the outer environment. So for tests that install Python packages we need
+to make sure those are installed in a test environment. This means patching the inmanta.env.process_env to be an interface
+to a test environment instead of the outer environment.
+The following fixtures manage test environments:
+- snippetcompiler_clean: activates the Project's compiler venv and patches inmanta.env.process_env to use this same venv
+    as if it were the outer venv. The activation and patch is applied when compiling the first snippet.
+- snippetcompiler: same as snippetcompiler_clean but the compiler venv is shared among all functions using the
```
Comment (arnaudsjs) [inline tests/conftest.py:47]:
```suggestion
- snippetcompiler: same as snippetcompiler_clean but the compiler venv is shared among all tests using the
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 2 above)
Comment (sanderr) [reply tests/conftest.py:49]: That's incorrect, thanks. I've updated the docstring.

**Pull request closed without merging**